cmd/nse/nse.go

```diff
@@ -0,0 +1,171 @@
+	"context"
+	"flag"
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"time"
+
+	"path/filepath"
+
+	"github.com/ligato/networkservicemesh/netmesh/model/netmesh"
+	"github.com/ligato/networkservicemesh/pkg/nsm/apis/common"
+	"github.com/ligato/networkservicemesh/pkg/nsm/apis/nsmconnect"
+	"github.com/ligato/networkservicemesh/plugins/nsmserver"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+const (
+	// clientConnectionTimeout defines time the client waits for establishing connection with the server
+	clientConnectionTimeout = time.Second * 60
+	// clientConnectionTimeout defines retry interval for establishing connection with the server
+	clientConnectionRetry = time.Second * 2
+)
+
+var (
+	clientSocketPath     = path.Join(nsmserver.SocketBaseDir, nsmserver.ServerSock)
+	clientSocketUserPath = flag.String("nsm-socket", "", "Location of NSM process client access socket")
+	nseSocketName        = flag.String("nse-socket", "nse.ligato.io.sock", "Name of NSE socket whcih will be used by NSM for Connection Request call")
+)
+
+func dial(ctx context.Context, unixSocketPath string) (*grpc.ClientConn, error) {
+	c, err := grpc.DialContext(ctx, unixSocketPath, grpc.WithInsecure(), grpc.WithBlock(),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}),
+	)
+
+	return c, err
+}
+
+func main() {
+	flag.Parse()
+	flag.Set("logtostderr", "true")
+
+	// Checking if nsm client socket exists and of not crash NSE
+	clientSocket := clientSocketPath
+	if clientSocketUserPath != nil {
+		clientSocket = *clientSocketUserPath
+	}
+
+	if _, err := os.Stat(clientSocket); err != nil {
+		logrus.Errorf("nse: failure to access nsm socket at %s with error: %+v, exiting...", clientSocket, err)
+		os.Exit(1)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), clientConnectionTimeout)
+	defer cancel()
+	conn, err := dial(ctx, clientSocket)
+	if err != nil {
+		logrus.Errorf("nse: failure to communicate with the socket %s with error: %+v", clientSocket, err)
+		os.Exit(1)
+	}
+	defer conn.Close()
+	logrus.Infof("nse: connection to nsm server on socket: %s succeeded.", clientSocket)
+
+	// NSM socket path will be used to drop NSE socket for NSM's Connection request
+	nsePath, _ := filepath.Split(clientSocket)
+	if err := socketCleanup(path.Join(nsePath, *nseSocketName)); err != nil {
+		logrus.Errorf("nse: failure to cleanup stale sicket %s with error: %+v", path.Join(nsePath, *nseSocketName), err)
+		os.Exit(1)
+	}
+	nse, err := net.Listen("unix", path.Join(nsePath, *nseSocketName))
+	grpcServer := grpc.NewServer([]grpc.ServerOption{}...)
+
+	go func() {
+		if err := grpcServer.Serve(nse); err != nil {
+			logrus.Errorf("nse: failed to start grpc server on socket %s with error: %+v ", path.Join(nsePath, *nseSocketName), err)
+		}
+	}()
+	// Check if the socket of device plugin server is operation
+	if err := socketOperationCheck(path.Join(nsePath, *nseSocketName)); err != nil {
```

Comment:
does grpcServer.Server need to start before socketOperationCheck runs? We may have a race condition here.

Reply:
grpc Server starts right before this call, then this calls check is the socket is `alive`. There is a 5 seconds timeout, which gives the server enough time to come up.